### PR TITLE
Bug/makes all delegates weak

### PR DESCRIPTION
**Pull request by benRmc**
**Description**

**Issues that should be CLOSED by merge of this PR:**
* Fixes #

**Related issues that should be LINKED to this PR:**
* Connects #

### CODE REVIEW

File: Traveler/Storyboards/Base.lproj/Main.storyboard [hunk 1]
```diff
@@ -3,7 +3,7 @@
     <device id="retina5_9" orientation="portrait" appearance="light"/>
     <dependencies>
         <deployment identifier="iOS"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15509"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15510"/>
```
Comment (rhoque-glx) [inline Traveler/Storyboards/Base.lproj/Main.storyboard]: Why was this file modified?

File: traveler-ios-ui/TravelerKitUI/Models/ParkingFilterContext.swift [hunk 2]
```diff
@@ -15,21 +15,37 @@ public protocol ParkingFilterContextObserving: class {
...
 
     var selectedFilter: Int? {
         didSet {
-            observers.forEach { $0.parkingFilterContextDidChangeSelectedFilter(self)}
+            notifyObserver()
         }
     }
 
-    private var observers = [ParkingFilterContextObserving]()
+    private struct Observation {
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Models/ParkingFilterContext.swift]: The observer logic should be abstracted.  Lots of repetition where only the Observing type differs.  Make one generic Observing type and utilize in specific instances.

File: traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard [hunk 3]
```diff
@@ -3,7 +3,7 @@
     <device id="retina6_1" orientation="portrait" appearance="light"/>
     <dependencies>
         <deployment identifier="iOS"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15509"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15510"/>
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard]: Why did this file change?

File: traveler-ios-ui/TravelerKitUI/Storyboards/Product.storyboard [hunk 4]
```diff
@@ -3,7 +3,7 @@
     <device id="retina6_1" orientation="portrait" appearance="light"/>
     <dependencies>
         <deployment identifier="iOS"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15509"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="15510"/>
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Storyboards/Product.storyboard]: Same?

File: traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingQueryFormViewController.swift [hunk 5]
```diff
@@ -39,8 +39,8 @@ public class ParkingQueryFormViewController: UIViewController {
     public override func viewDidAppear(_ animated: Bool) {
         parkingFilterContext?.addObserver(self)
     }
-
-    public override func viewWillDisappear(_ animated: Bool) {
+    
+    deinit {
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingQueryFormViewController.swift]: Added in `viewDidAppear`, remove in `viewWillDisappear`.  Otherwise, need to add in `viewDidLoad` and remove in `deinit`

File: traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingResultMapListController.swift [hunk 6]
```diff
@@ -1,5 +1,5 @@
 //
-//  ParkingQueryViewController.swift
+//  ParkingResultMapListViewController.swift
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingResultMapListController.swift]: `ParkingResultMapViewController` makes more sense

File: traveler-swift-core/TravelerKit/Models/ErrorContext.swift [hunk 7]
```diff
@@ -35,29 +38,36 @@ public class ErrorContext {
...
-        _ = observers.firstIndex(where: { observer === $0 }).flatMap({ observers.remove(at: $0) })
+        let id = ObjectIdentifier(observer)
+        observations.removeValue(forKey: id)
     }
 
     private func notifyObservers() {
-        for observer in observers {
+        for (id, observation) in observations {
+            guard let observer = observation.observer else {
+                observations.removeValue(forKey: id)
```
Comment (rhoque-glx) [inline traveler-swift-core/TravelerKit/Models/ErrorContext.swift]: Why are observers removed once notified?  Is this expected behaviour?

Comment (benRmc) [reply traveler-swift-core/TravelerKit/Models/ErrorContext.swift]: It's only removed if observation.observer is nil.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingResultMapListController.swift (hunk 6 above)
Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/ParkingResultMapListController.swift]: This is the container of both ParkingResultMapViewController and ParkingResultListViewController.

File: traveler-ios-ui/TravelerKitUI/Storyboards/Product.storyboard (hunk 4 above)
Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/Storyboards/Product.storyboard]: It's because of the segue change I think.

Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/Storyboards/Product.storyboard]: It's because of the segue change I think.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/BookingOptionsViewController.swift [hunk 8]
```diff
@@ -14,6 +14,8 @@ public protocol BookingOptionsViewControllerDelegate: class {
     func bookingOptionsViewController(_ controller: BookingOptionsViewController, didFinishWith purchaseForm: PurchaseForm)
 }
 
+let optionCellIdentifier = "optionCellIdentifier"
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/BookingOptionsViewController.swift]: move into class definition, doesn't need to be global

Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/BookingOptionsViewController.swift]: I think declaring it globally is a pattern we follow in case it's used somewhere else in the future. Check `stringCellIdentifier` which is also used in its own class but also declared globally.

Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/BookingOptionsViewController.swift]: Its not good practice to do that. It will make the code harder to maintain a and follow if you're using identifiers from different files. For the moment, add it to the class scope and in the near future we will refactor our identifiers into a single struct.

File: traveler-ios-ui/TravelerKitUI/Models/ParkingResultContext.swift [hunk 9]
```diff
@@ -14,38 +14,45 @@ public protocol ParkingResultContextObserving: class {
...
-public class ParkingResultContext {
+public class ParkingResultContext: ObservingContext {
     private(set) var spots: [ParkingSpot]?
 
     var selectedIndex: Int? {
         didSet {
-            observers.forEach {
-                $0.parkingResultContextDidChangeSelectedIndex(self)
-            }
+            notifyObserversWithSelectedIndexChange()
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/Models/ParkingResultContext.swift]: You're doing notifyObserversWithResultUpdate on a background thread but not here, why?.

File: traveler-swift-core/TravelerKit/Models/ErrorContext.swift [hunk 10]
```diff
@@ -1,64 +0,0 @@
-//
-//  ErrorContext.swift
-//  TravelerKit
-//
-//  Created by Ata Namvari on 2018-12-05.
-//  Copyright © 2018 Ata Namvari. All rights reserved.
```
Comment (OmarPadierna) [inline traveler-swift-core/TravelerKit/Models/ErrorContext.swift]: Why did you delete Error context?. What happened to the entities that use it?.
